File: src/Exporter.NX.Model/SketchPlaneMath.cs
// SPDX-License-Identifier: GPL-2.0-only
using System;
using System.Collections.Generic;

namespace Oblikovati.Exporter.NX.Model
{
    /// <summary>A fitted sketch plane frame: an origin and orthonormal in-plane axes (model space).</summary>
    public sealed class SketchPlaneFrame
    {
        public double[] Origin { get; set; } = { 0, 0, 0 };

        public double[] XAxis { get; set; } = { 1, 0, 0 };

        public double[] YAxis { get; set; } = { 0, 1, 0 };

        /// <summary>Projects a model-space point onto the frame's 2D (u, v) coordinates.</summary>
        public double[] To2D(double[] p)
        {
            double[] d = SketchPlaneMath.Sub(p, Origin);
            return new[] { SketchPlaneMath.Dot(d, XAxis), SketchPlaneMath.Dot(d, YAxis) };
        }
    }

    /// <summary>
    /// Pure geometry for sketch extraction: fitting a plane frame to the 3D points of an NX
    /// sketch's curves and projecting points into it. Kept NXOpen-free so it is unit-testable
    /// (the surrounding extraction that reads NXOpen curves is not). Robust enough for the
    /// planar sketches NX produces; a non-planar point set fits to its first valid triangle.
    /// </summary>
    public static class SketchPlaneMath
    {
        /// <summary>
        /// Fits a plane frame to points: origin at their centroid, normal from the first
        /// non-degenerate triangle, X axis toward the farthest point, Y = normal x X. Falls
        /// back to the world XY frame when the points are collinear/degenerate.
        /// </summary>
        public static SketchPlaneFrame Fit(IReadOnlyList<double[]> points)
        {
            double[] origin = GeometryMath.Average(points);
            double[]? normal = FirstNormal(origin, points);
            if (normal == null)
            {
                return new SketchPlaneFrame { Origin = origin };
            }

            double[] x = FarthestInPlane(origin, normal, points);
            double[] y = Normalize(Cross(normal, x));
            return new SketchPlaneFrame { Origin = origin, XAxis = x, YAxis = y };
        }

        // The unit normal of the first non-collinear triple about the origin, or null.
        private static double[]? FirstNormal(double[] origin, IReadOnlyList<double[]> points)
        {
            for (int i = 0; i < points.Count; i++)
            {
                for (int j = i + 1; j < points.Count; j++)
                {
                    double[] n = Cross(Sub(points[i], origin), Sub(points[j], origin));
                    if (Length(n) > 1e-9)
                    {
                        return Normalize(n);
                    }
                }
            }

            return null;
        }

        // A unit in-plane X axis toward the point farthest from the origin (projected into the plane).
        private static double[] FarthestInPlane(double[] origin, double[] normal, IReadOnlyList<double[]> points)
        {
            double[] best = { 1, 0, 0 };
            double bestLen = 0;
            foreach (double[] p in points)
            {
                double[] d = Sub(p, origin);
                double[] inPlane = Sub(d, Scale(normal, Dot(d, normal)));
                double len = Length(inPlane);
                if (len > bestLen)
                {
                    bestLen = len;
                    best = inPlane;
                }
            }

            return bestLen > 1e-9 ? Normalize(best) : AnyPerpendicular(normal);
        }

        private static double[] AnyPerpendicular(double[] n)
        {
            double[] seed = Math.Abs(n[0]) < 0.9 ? new double[] { 1, 0, 0 } : new double[] { 0, 1, 0 };
            return Normalize(Cross(n, seed));
        }

        public static double[] Sub(double[] a, double[] b) => new[] { a[0] - b[0], a[1] - b[1], a[2] - b[2] };

        public static double Dot(double[] a, double[] b) => a[0] * b[0] + a[1] * b[1] + a[2] * b[2];

        public static double[] Cross(double[] a, double[] b) => new[]
        {
            a[1] * b[2] - a[2] * b[1],
            a[2] * b[0] - a[0] * b[2],
            a[0] * b[1] - a[1] * b[0],
        };

        public static double[] Scale(double[] a, double s) => new[] { a[0] * s, a[1] * s, a[2] * s };

        public static double Length(double[] a) => Math.Sqrt(Dot(a, a));

        public static double[] Normalize(double[] a)
        {
            double len = Length(a);
            return len == 0 ? new double[] { 0, 0, 0 } : Scale(a, 1 / len);
        }
    }
}

File: python/oblikovati_exporter_nx/model/sketch_plane_math.py
# SPDX-License-Identifier: GPL-2.0-only
"""Pure geometry for sketch extraction: fit a plane frame to 3D points, project into it.

Kept NXOpen-free so it is unit-testable (the surrounding extraction that reads NXOpen
curves is not). Robust enough for the planar sketches NX produces; a non-planar point
set fits to its first valid triangle.
"""
from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import List, Optional, Sequence

from . import geometry_math


def sub(a: Sequence[float], b: Sequence[float]) -> List[float]:
    return [a[0] - b[0], a[1] - b[1], a[2] - b[2]]


def dot(a: Sequence[float], b: Sequence[float]) -> float:
    return a[0] * b[0] + a[1] * b[1] + a[2] * b[2]


def cross(a: Sequence[float], b: Sequence[float]) -> List[float]:
    return [
        a[1] * b[2] - a[2] * b[1],
        a[2] * b[0] - a[0] * b[2],
        a[0] * b[1] - a[1] * b[0],
    ]


def scale(a: Sequence[float], s: float) -> List[float]:
    return [a[0] * s, a[1] * s, a[2] * s]


def length(a: Sequence[float]) -> float:
    return math.sqrt(dot(a, a))


def normalize(a: Sequence[float]) -> List[float]:
    ln = length(a)
    return [0.0, 0.0, 0.0] if ln == 0 else scale(a, 1 / ln)


@dataclass
class SketchPlaneFrame:
    """A fitted sketch plane frame: origin + orthonormal in-plane axes (model space)."""

    origin: List[float] = field(default_factory=lambda: [0.0, 0.0, 0.0])
    xaxis: List[float] = field(default_factory=lambda: [1.0, 0.0, 0.0])
    yaxis: List[float] = field(default_factory=lambda: [0.0, 1.0, 0.0])

    def to_2d(self, p: Sequence[float]) -> List[float]:
        """Projects a model-space point onto the frame's 2D (u, v) coordinates."""
        d = sub(p, self.origin)
        return [dot(d, self.xaxis), dot(d, self.yaxis)]


def fit(points: Sequence[Sequence[float]]) -> SketchPlaneFrame:
    """Fit a plane frame to points.

    Origin at their centroid, normal from the first non-degenerate triangle, X axis
    toward the farthest in-plane point, Y = normal x X. Falls back to the world XY frame
    when the points are collinear/degenerate.
    """
    origin = geometry_math.average(points)
    normal = _first_normal(origin, points)
    if normal is None:
        return SketchPlaneFrame(origin=origin)
    x = _farthest_in_plane(origin, normal, points)
    y = normalize(cross(normal, x))
    return SketchPlaneFrame(origin=origin, xaxis=x, yaxis=y)


def _first_normal(origin: Sequence[float], points: Sequence[Sequence[float]]) -> Optional[List[float]]:
    for i in range(len(points)):
        for j in range(i + 1, len(points)):
            n = cross(sub(points[i], origin), sub(points[j], origin))
            if length(n) > 1e-9:
                return normalize(n)
    return None


def _farthest_in_plane(
    origin: Sequence[float], normal: Sequence[float], points: Sequence[Sequence[float]]
) -> List[float]:
    best = [1.0, 0.0, 0.0]
    best_len = 0.0
    for p in points:
        d = sub(p, origin)
        in_plane = sub(d, scale(normal, dot(d, normal)))
        ln = length(in_plane)
        if ln > best_len:
            best_len = ln
            best = in_plane
    return normalize(best) if best_len > 1e-9 else _any_perpendicular(normal)


def _any_perpendicular(n: Sequence[float]) -> List[float]:
    seed = [1.0, 0.0, 0.0] if abs(n[0]) < 0.9 else [0.0, 1.0, 0.0]
    return normalize(cross(n, seed))
